modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/AssignmentsTracker.java

```diff
@@ -156,6 +184,124 @@ public CompletableFuture<List<TokenizedAssignments>> getAssignments(
                 }));
     }
 
+    @Override
+    public CompletableFuture<List<TokenizedAssignments>> awaitNonEmptyAssignments(
+            List<? extends ReplicationGroupId> replicationGroupIds,
+            HybridTimestamp clusterTimeToAwait,
+            long timeoutMillis
+    ) {
+        return msManager
+                .clusterTime()
+                .waitFor(clusterTimeToAwait)
+                .thenCompose(ignored -> inBusyLock(busyLock, () -> {
+                    long now = coarseCurrentTimeMillis();
+                    return awaitNonEmptyAssignmentsWithCheckMostRecent(replicationGroupIds, now, timeoutMillis);
+                }))
+                .thenApply(identity());
+    }
+
+    private CompletableFuture<List<TokenizedAssignments>> awaitNonEmptyAssignmentsWithCheckMostRecent(
+            List<? extends ReplicationGroupId> replicationGroupIds,
+            long startTime,
+            long timeoutMillis
+    ) {
+        Map<ReplicationGroupId, TokenizedAssignments> assignmentsMap = stableAssignments();
+
+        Map<Integer, CompletableFuture<TokenizedAssignments>> futures = new HashMap<>();
+        List<TokenizedAssignments> result = new ArrayList<>(replicationGroupIds.size());
+
+        for (int i = 0; i < replicationGroupIds.size(); i++) {
+            ReplicationGroupId groupId = replicationGroupIds.get(i);
+
+            TokenizedAssignments a = assignmentsMap.get(groupId);
+            result.add(a);
+
+            if (a.nodes().isEmpty()) {
+                if (timeoutMillis > 0) {
+                    futures.put(i, nonEmptyAssignmentFuture(groupId, timeoutMillis));
+                } else {
+                    // If timeout is zero or less, then this group is failed, the correct exception will be thrown
+                    // in #checkEmptyAssignmentsReasons().
+                    futures.put(i, failedFuture(new TimeoutException()));
+                }
+            }
+        }
+
+        if (futures.isEmpty()) {
+            return completedFuture(result);
+        } else {
+            return allOf(futures.values())
+                    .handle((unused, ex) -> {
+                        if (ex == null) {
+                            // Get the most recent assignments after the waiting.
+                            long now = System.currentTimeMillis();
+                            long newTimeoutMillis = timeoutMillis - (now - startTime);
+                            return awaitNonEmptyAssignmentsWithCheckMostRecent(replicationGroupIds, startTime, newTimeoutMillis);
+                        } else {
+                            return checkEmptyAssignmentsReasons(replicationGroupIds, futures, ex);
+                        }
+                    })
+                    .thenCompose(identity());
+        }
+    }
+
+    private CompletableFuture<List<TokenizedAssignments>> checkEmptyAssignmentsReasons(
```

Comment:
I guess it's a bit excessive. allOf() will be completed exceptionally on the very first exceptionally completed future, thus you may just unwrap the cause instead of iterating over all futures in assignmentFuturesMap.

Reply:
this method not only unwraps exception but transforms it to most relevant, i.e. EmptyDataNodesException